### DISCUSSION TITLE
how to bundle firebase and firebaseui

### DISCUSSION THREAD

**Discussion by lhk** [parcel-bundler/parcel#3519] [General]
# ❔ Question

How can I bundle Firebase and FirebaseUI, installed with npm, with Parcel ?

## 🔦 Context

My webapp uses FirebaseUI to let users log in. The installation instructions list a variety of options to install FirebaseUI: https://github.com/firebase/firebaseui-web#installation

- by adding firebase's CDN and adding <script> tags to the HTML
- by installing and importing npm modules

The first option works fine for me. But CDN imports in the HTML are messy, I would like to use parcel to bundle the imports instead.
As soon as I add the import statements to my typescript code, FirebaseUI stops working and complains that `firebaseui.auth` is undefined. It works only, if there are only the CDN script tags in the HTML.

## 💻 Code Sample

To run the following code, you have to enter a correct Firebase config in the typescript code and bundle **index.html** with parcel. Please note, you can't just open the compiled html in your browser. Firebase expects a hosted environment.
```
npm install -g serve
serve dist/
```

**index.html**:

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>Sample FirebaseUI App</title>

    <script src="https://www.gstatic.com/firebasejs/6.6.1/firebase-app.js"></script>
    <script src="https://www.gstatic.com/firebasejs/6.6.1/firebase-auth.js"></script>
    <script src="https://cdn.firebase.com/libs/firebaseui/4.2.0/firebaseui.js"></script>
    <link type="text/css" rel="stylesheet" href="https://cdn.firebase.com/libs/firebaseui/4.2.0/firebaseui.css" />

    <script type="text/javascript" src="index.ts"></script>
  </head>
  <body>
    <h1>Welcome to My Awesome App</h1>
    <div id="firebaseui-auth-container"></div>
  </body>
</html>
```

**index.ts**:
```

// TODO: remove these imports to make it work
import firebase from "firebase"
import "firebase/auth"
import firebaseui from "firebaseui"

window.onload = () => {
    // this is defined
    console.log("firebase.auth", firebase.auth);

    // this is undefined
    console.log("firebaseui.auth", firebaseui.auth);

    // configure firebase
    var config = {
          // TODO: add your Firebase web app credentials to make it work
      };

    firebase.initializeApp(config);

    // code crashes here
    var ui = new firebaseui.auth.AuthUI(firebase.auth())

    var uiConfig = {
        callbacks: {
            signInSuccessWithAuthResult: function (authResult, redirectUrl) {
                return false;
            }
        },
        signInFlow: 'popup',
        signInOptions: [
            firebase.auth.EmailAuthProvider.PROVIDER_ID
        ]
    };
    ui.start('#firebaseui-auth-container', uiConfig);
}
```

**package.json**:
```
{
  "name": "temp",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "firebase": "^6.6.1",
    "firebaseui": "^4.2.0",
    "serve": "^11.1.0"
  },
  "devDependencies": {
    "typescript": "^3.6.3"
  }
}

```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.12.3
| Node             | v12.10.0
| npm/Yarn         | 6.10.3
| Operating System | 64bit Pop!_os

**Comment by mischnic** [comment]
This should work 🤷‍♂ 

```js
import * as firebase from "firebase/app";
import "firebase/auth";
import * as firebaseui from "firebaseui";
```

**Comment by lhk** [comment]
@mischnic did you try it, does it work for you ?

**Comment by mischnic** [comment]
According to the documentation, it should. I've looked at the published `firebaseui` npm package and this is the import syntax that is needed. Don't know if this is firebase's fault or Parcel does something wrong.

**Comment by dawogfather** [comment]
Same issue here. Did you get it resolved? I was using the above import syntax exactly and it doesn’t work.

**Comment by mischnic** [comment]
(Maybe related to #3387?)

**Comment by lhk** [comment]
@dawogfather, I haven't gotten it resolved yet :(
As a workaround I'm importing the dependencies in html. Not maintainable at all, but currently seems like the easiest way.

**Comment by lhk** [comment]
@mischnic , the linked issue does indeed look very similar. Fingers crossed that the PR fixes 

Do you know the correct procedure here?
Do I close the issue and wait for results from #3387, to maybe reopen it?
Or should I leave it open for now ?

**Comment by mischnic** [comment]
Let's keep both open until the next release with that PR and then test both issues.

**Comment by lhk** [comment]
not stale

**Comment by ninest** [comment]
Is there a full example we can refer to? I found [this one](https://github.com/codepiranhas/firebase-parcel-example), but I'm not sure if it's the best because the config is directly in index.html, and they're not using the packages for firebase

**Reply by mischnic** [reply]
Does this help? https://github.com/parcel-bundler/parcel-benchmark-action/blob/master/benchmarks/react-hn/src/services/HNService.js

**Comment by ninest** [comment]
@mischnic Thanks for the help, but I figured it out. Not sure if this is the best way, but I made file called `firebase.js`:

```js
import firebase from 'firebase/app';
import 'firebase/auth';

const firebaseConfig = {
  // config here
};
export const app = firebase.initializeApp(firebaseConfig);
```

Then whenever I need to use firebase, I can import in my script like this:

```js
import { app } from './firebase.js';
import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

...

const db = firebase.firestore(app);
```

**Comment by ksmigiel** [comment]
I am also having problem with Parcel and firebase/firebaseui. CDN of course works, but when using npm modules I cannot import neither firebase or firebaseui successfully.

**Comment by ksmigiel** [comment]
Ok after some research I have managed to have working imports in Parcel by importing manually CommonJS instead of ES6 modules:
```js
import firebase from "firebase/app/dist/index.cjs";
import "firebase/auth/dist/index.cjs";

import firebaseui from "firebaseui/dist/npm";
import "firebaseui/dist/firebaseui.css";
```
Anyone from Parcel Team could look into this why we need to do this? Is that connected with fact that firebase is exported using Google Closure Compiler?

**Reply by mischnic** [reply]
I'm not familiar with firebase, could you provide a JS sample that should work but throws an error?
